File: backend/src/main/java/com/tuorg/materials/service/MaterialService.java
package com.tuorg.materials.service;

import com.tuorg.materials.dto.MaterialCreateDto;
import com.tuorg.materials.dto.MaterialDto;

import java.time.LocalDate;
import java.util.List;

public interface MaterialService {
    List<MaterialDto> getAll();
    List<MaterialDto> getByTipo(String tipo);
    List<MaterialDto> getByFechaCompra(LocalDate fecha);
    List<MaterialDto> getByCiudad(Long ciudadId);
    MaterialDto create(MaterialCreateDto dto);
    MaterialDto update(Long id, MaterialCreateDto dto);
}

File: frontend/src/app/material.service.ts
import { HttpClient } from '@angular/common/http';
import { Injectable } from '@angular/core';
@Injectable({ providedIn: 'root' })
export class MaterialService {
  private base = '/api/materials';
  constructor(private http: HttpClient) {}
  getAll() { return this.http.get<any>(this.base); }
  getByTipo(tipo: string) { return this.http.get<any>(this.base + '?tipo=' + tipo); }
  getByFechaCompra(fecha: string) { return this.http.get<any>(this.base + '?fechaCompra=' + fecha); }
  getByCity(cityId: number) { return this.http.get<any>(this.base + '/city/' + cityId); }
  create(payload: any) { return this.http.post<any>(this.base, payload); }
  update(id: number, payload: any) { return this.http.put<any>(this.base + '/' + id, payload); }
}
